// molstar/lib/commonjs/mol-geo/geometry/animation.js
"use strict";
/**
 * Copyright (c) 2026 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Alexander Rose <alexander.rose@weirdbyte.de>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getAnimationParam = getAnimationParam;
exports.areAnimationPropsEqual = areAnimationPropsEqual;
exports.createAnimationValues = createAnimationValues;
exports.updateAnimationValues = updateAnimationValues;
const mol_util_1 = require("../../mol-util/index.js");
const param_definition_1 = require("../../mol-util/param-definition.js");
function getAnimationParam() {
    return param_definition_1.ParamDefinition.Group({
        wiggleMode: param_definition_1.ParamDefinition.Select('position', [['position', 'Position'], ['group', 'Group']], { description: 'Noise seeding mode. Position: spatially correlated (nearby atoms move together). Group: per-group independent noise.' }),
        wiggleSpeed: param_definition_1.ParamDefinition.Numeric(7, { min: 0, max: 10, step: 0.1 }, { description: 'Speed of vertex wiggle animation.' }),
        wiggleAmplitude: param_definition_1.ParamDefinition.Numeric(0, { min: 0, max: 5, step: 0.01 }, { description: 'Amplitude of vertex wiggle animation.' }),
        wiggleFrequency: param_definition_1.ParamDefinition.Numeric(0.2, { min: 0.01, max: 2, step: 0.01 }, { description: 'Spatial frequency of vertex wiggle noise (position mode). Lower values correlate nearby atoms more.' }),
        tumbleSpeed: param_definition_1.ParamDefinition.Numeric(1, { min: 0, max: 10, step: 0.1 }, { description: 'Speed of instance tumble animation.' }),
        tumbleAmplitude: param_definition_1.ParamDefinition.Numeric(0, { min: 0, max: 10, step: 0.1 }, { description: 'Amplitude of instance tumble animation. In Ångströms of implied surface displacement.' }),
        tumbleFrequency: param_definition_1.ParamDefinition.Numeric(0.2, { min: 0, max: 2, step: 0.01 }, { description: 'Spatial frequency multiplier for tumble noise.' }),
    });
}
function areAnimationPropsEqual(a, b) {
    return a.wiggleMode === b.wiggleMode
        && a.wiggleSpeed === b.wiggleSpeed
        && a.wiggleAmplitude === b.wiggleAmplitude
        && a.wiggleFrequency === b.wiggleFrequency
        && a.tumbleSpeed === b.tumbleSpeed
        && a.tumbleAmplitude === b.tumbleAmplitude
        && a.tumbleFrequency === b.tumbleFrequency;
}
function createAnimationValues(props) {
    return {
        uWiggleSpeed: mol_util_1.ValueCell.create(props.wiggleSpeed),
        uWiggleAmplitude: mol_util_1.ValueCell.create(props.wiggleAmplitude),
        uWiggleFrequency: mol_util_1.ValueCell.create(props.wiggleFrequency),
        uWiggleMode: mol_util_1.ValueCell.create(props.wiggleMode === 'position' ? 0 : 1),
        uTumbleSpeed: mol_util_1.ValueCell.create(props.tumbleSpeed),
        uTumbleAmplitude: mol_util_1.ValueCell.create(props.tumbleAmplitude),
        uTumbleFrequency: mol_util_1.ValueCell.create(props.tumbleFrequency),
    };
}
function updateAnimationValues(values, props) {
    mol_util_1.ValueCell.updateIfChanged(values.uWiggleSpeed, props.wiggleSpeed);
    mol_util_1.ValueCell.updateIfChanged(values.uWiggleAmplitude, props.wiggleAmplitude);
    mol_util_1.ValueCell.updateIfChanged(values.uWiggleFrequency, props.wiggleFrequency);
    mol_util_1.ValueCell.updateIfChanged(values.uWiggleMode, props.wiggleMode === 'position' ? 0 : 1);
    mol_util_1.ValueCell.updateIfChanged(values.uTumbleSpeed, props.tumbleSpeed);
    mol_util_1.ValueCell.updateIfChanged(values.uTumbleAmplitude, props.tumbleAmplitude);
    mol_util_1.ValueCell.updateIfChanged(values.uTumbleFrequency, props.tumbleFrequency);
}
