// molstar/lib/commonjs/mol-geo/geometry/wiggle-data.js
"use strict";
/**
 * Copyright (c) 2026 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Alexander Rose <alexander.rose@weirdbyte.de>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.applyWiggleValue = applyWiggleValue;
exports.getWiggleAverage = getWiggleAverage;
exports.clearWiggle = clearWiggle;
exports.createWiggle = createWiggle;
exports.createEmptyWiggle = createEmptyWiggle;
const value_cell_1 = require("../../mol-util/value-cell.js");
const linear_algebra_1 = require("../../mol-math/linear-algebra.js");
const util_1 = require("../../mol-gl/renderable/util.js");
function applyWiggleValue(array, start, end, value) {
    for (let i = start; i < end; ++i) {
        array[i] = value * 255;
    }
    return true;
}
function getWiggleAverage(array, count) {
    if (count === 0 || array.length < count)
        return 0;
    let sum = 0;
    for (let i = 0; i < count; ++i) {
        sum += array[i];
    }
    return sum / (255 * count);
}
function clearWiggle(array, start, end) {
    array.fill(0, start, end);
}
function createWiggle(count, type, wiggleData) {
    const wiggle = (0, util_1.createTextureImage)(Math.max(1, count), 1, Uint8Array, wiggleData && wiggleData.tWiggle.ref.value.array);
    if (wiggleData) {
        value_cell_1.ValueCell.update(wiggleData.tWiggle, wiggle);
        value_cell_1.ValueCell.update(wiggleData.uWiggleTexDim, linear_algebra_1.Vec2.create(wiggle.width, wiggle.height));
        value_cell_1.ValueCell.updateIfChanged(wiggleData.dWiggle, count > 0);
        value_cell_1.ValueCell.updateIfChanged(wiggleData.wiggleAverage, getWiggleAverage(wiggle.array, count));
        value_cell_1.ValueCell.updateIfChanged(wiggleData.dWiggleType, type);
        return wiggleData;
    }
    else {
        return {
            tWiggle: value_cell_1.ValueCell.create(wiggle),
            uWiggleTexDim: value_cell_1.ValueCell.create(linear_algebra_1.Vec2.create(wiggle.width, wiggle.height)),
            dWiggle: value_cell_1.ValueCell.create(count > 0),
            wiggleAverage: value_cell_1.ValueCell.create(0),
            dWiggleType: value_cell_1.ValueCell.create(type),
            uWiggleStrength: value_cell_1.ValueCell.create(1),
        };
    }
}
const emptyWiggleTexture = { array: new Uint8Array(1), width: 1, height: 1 };
function createEmptyWiggle(wiggleData) {
    if (wiggleData) {
        value_cell_1.ValueCell.update(wiggleData.tWiggle, emptyWiggleTexture);
        value_cell_1.ValueCell.update(wiggleData.uWiggleTexDim, linear_algebra_1.Vec2.create(1, 1));
        return wiggleData;
    }
    else {
        return {
            tWiggle: value_cell_1.ValueCell.create(emptyWiggleTexture),
            uWiggleTexDim: value_cell_1.ValueCell.create(linear_algebra_1.Vec2.create(1, 1)),
            dWiggle: value_cell_1.ValueCell.create(false),
            wiggleAverage: value_cell_1.ValueCell.create(0),
            dWiggleType: value_cell_1.ValueCell.create('groupInstance'),
            uWiggleStrength: value_cell_1.ValueCell.create(1),
        };
    }
}
