// molstar/lib/commonjs/mol-theme/wiggle.js
"use strict";
/**
 * Copyright (c) 2026 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Alexander Rose <alexander.rose@weirdbyte.de>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Wiggle = Wiggle;
const loci_1 = require("../mol-model/loci.js");
const structure_1 = require("../mol-model/structure.js");
const script_1 = require("../mol-script/script.js");
function Wiggle(kind, layers) {
    return { kind, layers };
}
(function (Wiggle) {
    Wiggle.Empty = { kind: 'empty-loci', layers: [] };
    function areEqual(wA, wB) {
        if (wA.layers.length === 0 && wB.layers.length === 0)
            return true;
        if (wA.layers.length !== wB.layers.length)
            return false;
        for (let i = 0, il = wA.layers.length; i < il; ++i) {
            if (wA.layers[i].value !== wB.layers[i].value)
                return false;
            if (!loci_1.Loci.areEqual(wA.layers[i].loci, wB.layers[i].loci))
                return false;
        }
        return true;
    }
    Wiggle.areEqual = areEqual;
    function isEmpty(wiggle) {
        return wiggle.layers.length === 0;
    }
    Wiggle.isEmpty = isEmpty;
    function remap(wiggle, structure) {
        if (wiggle.kind === 'element-loci') {
            const layers = [];
            for (const layer of wiggle.layers) {
                let { loci, value } = layer;
                loci = structure_1.StructureElement.Loci.remap(loci, structure);
                if (!structure_1.StructureElement.Loci.isEmpty(loci)) {
                    layers.push({ loci, value });
                }
            }
            return { kind: 'element-loci', layers };
        }
        else {
            return wiggle;
        }
    }
    Wiggle.remap = remap;
    function merge(wiggle) {
        if (isEmpty(wiggle))
            return wiggle;
        if (wiggle.kind === 'element-loci') {
            const { structure } = wiggle.layers[0].loci;
            const map = new Map();
            let shadowed = structure_1.StructureElement.Loci.none(structure);
            for (let i = 0, il = wiggle.layers.length; i < il; ++i) {
                let { loci, value } = wiggle.layers[il - i - 1]; // process from end
                loci = structure_1.StructureElement.Loci.subtract(loci, shadowed);
                shadowed = structure_1.StructureElement.Loci.union(loci, shadowed);
                if (!structure_1.StructureElement.Loci.isEmpty(loci)) {
                    if (map.has(value)) {
                        loci = structure_1.StructureElement.Loci.union(loci, map.get(value));
                    }
                    map.set(value, loci);
                }
            }
            const layers = [];
            map.forEach((loci, value) => {
                layers.push({ loci, value });
            });
            return { kind: 'element-loci', layers };
        }
        else {
            return wiggle;
        }
    }
    Wiggle.merge = merge;
    function filter(wiggle, filter) {
        if (isEmpty(wiggle))
            return wiggle;
        if (wiggle.kind === 'element-loci') {
            const { structure } = wiggle.layers[0].loci;
            const layers = [];
            for (const layer of wiggle.layers) {
                let { loci, value } = layer;
                // filter by first map to the `filter` structure and
                // then map back to the original structure of the wiggle loci
                const filtered = structure_1.StructureElement.Loci.remap(loci, filter);
                loci = structure_1.StructureElement.Loci.remap(filtered, structure);
                if (!structure_1.StructureElement.Loci.isEmpty(loci)) {
                    layers.push({ loci, value });
                }
            }
            return { kind: 'element-loci', layers };
        }
        else {
            return wiggle;
        }
    }
    Wiggle.filter = filter;
    function ofScript(scriptLayers, structure) {
        const layers = [];
        for (let i = 0, il = scriptLayers.length; i < il; ++i) {
            const { script, value } = scriptLayers[i];
            const loci = script_1.Script.toLoci(script, structure);
            if (!structure_1.StructureElement.Loci.isEmpty(loci)) {
                layers.push({ loci, value });
            }
        }
        return { kind: 'element-loci', layers };
    }
    Wiggle.ofScript = ofScript;
    function ofBundle(bundleLayers, structure) {
        const layers = [];
        for (let i = 0, il = bundleLayers.length; i < il; ++i) {
            const { bundle, value } = bundleLayers[i];
            const loci = structure_1.StructureElement.Bundle.toLoci(bundle, structure.root);
            layers.push({ loci, value });
        }
        return { kind: 'element-loci', layers };
    }
    Wiggle.ofBundle = ofBundle;
    function toBundle(wiggle) {
        const layers = [];
        for (let i = 0, il = wiggle.layers.length; i < il; ++i) {
            const { loci, value } = wiggle.layers[i];
            const bundle = structure_1.StructureElement.Bundle.fromLoci(loci);
            layers.push({ bundle, value });
        }
        return { kind: 'element-loci', layers };
    }
    Wiggle.toBundle = toBundle;
})(Wiggle || (exports.Wiggle = Wiggle = {}));
