// molstar/lib/mol-geo/geometry/wiggle-data.js
/**
 * Copyright (c) 2026 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Alexander Rose <alexander.rose@weirdbyte.de>
 */
import { ValueCell } from '../../mol-util/value-cell.js';
import { Vec2 } from '../../mol-math/linear-algebra.js';
import { createTextureImage } from '../../mol-gl/renderable/util.js';
export function applyWiggleValue(array, start, end, value) {
    for (let i = start; i < end; ++i) {
        array[i] = value * 255;
    }
    return true;
}
export function getWiggleAverage(array, count) {
    if (count === 0 || array.length < count)
        return 0;
    let sum = 0;
    for (let i = 0; i < count; ++i) {
        sum += array[i];
    }
    return sum / (255 * count);
}
export function clearWiggle(array, start, end) {
    array.fill(0, start, end);
}
export function createWiggle(count, type, wiggleData) {
    const wiggle = createTextureImage(Math.max(1, count), 1, Uint8Array, wiggleData && wiggleData.tWiggle.ref.value.array);
    if (wiggleData) {
        ValueCell.update(wiggleData.tWiggle, wiggle);
        ValueCell.update(wiggleData.uWiggleTexDim, Vec2.create(wiggle.width, wiggle.height));
        ValueCell.updateIfChanged(wiggleData.dWiggle, count > 0);
        ValueCell.updateIfChanged(wiggleData.wiggleAverage, getWiggleAverage(wiggle.array, count));
        ValueCell.updateIfChanged(wiggleData.dWiggleType, type);
        return wiggleData;
    }
    else {
        return {
            tWiggle: ValueCell.create(wiggle),
            uWiggleTexDim: ValueCell.create(Vec2.create(wiggle.width, wiggle.height)),
            dWiggle: ValueCell.create(count > 0),
            wiggleAverage: ValueCell.create(0),
            dWiggleType: ValueCell.create(type),
            uWiggleStrength: ValueCell.create(1),
        };
    }
}
const emptyWiggleTexture = { array: new Uint8Array(1), width: 1, height: 1 };
export function createEmptyWiggle(wiggleData) {
    if (wiggleData) {
        ValueCell.update(wiggleData.tWiggle, emptyWiggleTexture);
        ValueCell.update(wiggleData.uWiggleTexDim, Vec2.create(1, 1));
        return wiggleData;
    }
    else {
        return {
            tWiggle: ValueCell.create(emptyWiggleTexture),
            uWiggleTexDim: ValueCell.create(Vec2.create(1, 1)),
            dWiggle: ValueCell.create(false),
            wiggleAverage: ValueCell.create(0),
            dWiggleType: ValueCell.create('groupInstance'),
            uWiggleStrength: ValueCell.create(1),
        };
    }
}
