// molstar/lib/commonjs/mol-plugin-ui/structure/procedural-animation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StructureProceduralAnimationControls = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
/**
 * Copyright (c) 2026 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Alexander Rose <alexander.rose@weirdbyte.de>
 */
const structure_wiggle_1 = require("../../mol-plugin-state/helpers/structure-wiggle.js");
const base_1 = require("../base.js");
const common_1 = require("../controls/common.js");
const icons_1 = require("../controls/icons.js");
class StructureProceduralAnimationControls extends base_1.CollapsableControls {
    defaultState() {
        return {
            isCollapsed: false,
            header: 'Procedural Animation',
            brand: { accent: 'gray', svg: icons_1.AnimationSvg }
        };
    }
    renderControls() {
        return (0, jsx_runtime_1.jsx)(StructureProceduralAnimation, {});
    }
}
exports.StructureProceduralAnimationControls = StructureProceduralAnimationControls;
class StructureProceduralAnimation extends base_1.PurePluginUIComponent {
    constructor() {
        super(...arguments);
        this.state = { busy: false };
    }
    get components() {
        return this.plugin.managers.structure.hierarchy.selection.structures.flatMap(s => s.components);
    }
    async applyUncertaintyWiggle() {
        this.setState({ busy: true });
        try {
            const options = this.plugin.managers.structure.component.state.options;
            await this.plugin.managers.structure.component.setOptions({
                ...options,
                animation: {
                    ...options.animation,
                    wiggleAmplitude: 0,
                    tumbleAmplitude: 0
                }
            });
            await (0, structure_wiggle_1.setStructureWiggleFromUncertainty)(this.plugin, this.components);
        }
        finally {
            this.setState({ busy: false });
        }
    }
    async applyDynamics() {
        this.setState({ busy: true });
        try {
            const options = this.plugin.managers.structure.component.state.options;
            await this.plugin.managers.structure.component.setOptions({
                ...options,
                animation: {
                    ...options.animation,
                    wiggleSpeed: 7,
                    wiggleAmplitude: 1,
                    wiggleFrequency: 0.2,
                }
            });
            await (0, structure_wiggle_1.clearStructureWiggle)(this.plugin, this.components);
        }
        finally {
            this.setState({ busy: false });
        }
    }
    async clearWiggle() {
        this.setState({ busy: true });
        try {
            const options = this.plugin.managers.structure.component.state.options;
            await this.plugin.managers.structure.component.setOptions({
                ...options,
                animation: {
                    ...options.animation,
                    wiggleAmplitude: 0,
                    tumbleAmplitude: 0
                }
            });
            await (0, structure_wiggle_1.clearStructureWiggle)(this.plugin, this.components);
        }
        finally {
            this.setState({ busy: false });
        }
    }
    render() {
        return (0, jsx_runtime_1.jsx)(jsx_runtime_1.Fragment, { children: (0, jsx_runtime_1.jsxs)("div", { className: 'msp-control-group-wrapper', children: [(0, jsx_runtime_1.jsx)("div", { className: 'msp-control-group-header', children: (0, jsx_runtime_1.jsx)("div", { children: (0, jsx_runtime_1.jsx)("b", { children: "Apply Wiggle" }) }) }), (0, jsx_runtime_1.jsxs)("div", { className: 'msp-flex-row', children: [(0, jsx_runtime_1.jsx)(common_1.Button, { title: 'Set wiggle speed to 8 and amplitude to 1', onClick: () => this.applyDynamics(), disabled: this.state.busy, children: "Dynamics" }), (0, jsx_runtime_1.jsx)(common_1.Button, { title: 'Set per-group wiggle amplitude based on B-factor / RMSF uncertainty', onClick: () => this.applyUncertaintyWiggle(), disabled: this.state.busy, children: "Uncertainty" }), (0, jsx_runtime_1.jsx)(common_1.Button, { title: 'Set wiggle/tumble amplitude to zero and remove per-group wiggle layers', onClick: () => this.clearWiggle(), disabled: this.state.busy, children: "Clear" })] })] }) });
    }
}
