// molstar/lib/commonjs/mol-plugin-state/animation/built-in/time.js
"use strict";
/**
 * Copyright (c) 2026 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Alexander Rose <alexander.rose@weirdbyte.de>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.AnimateTime = void 0;
const param_definition_1 = require("../../../mol-util/param-definition.js");
const model_1 = require("../model.js");
exports.AnimateTime = model_1.PluginStateAnimation.create({
    name: 'built-in.animate-time',
    display: { name: 'Animate Time', description: 'Animate the passage of time in the 3D scene' },
    isExportable: true,
    params: () => ({
        durationInMs: param_definition_1.ParamDefinition.Numeric(4000, { min: 100, max: 20000, step: 100 }),
    }),
    initialState: () => ({}),
    getDuration: p => ({ kind: 'fixed', durationMs: p.durationInMs }),
    async apply(animState, t, ctx) {
        return t.current < ctx.params.durationInMs
            ? { kind: 'next', state: animState }
            : { kind: 'finished' };
    }
});
