// molstar/lib/mol-plugin-ui/structure/procedural-animation.js
import { jsx as _jsx, jsxs as _jsxs, Fragment as _Fragment } from "react/jsx-runtime";
/**
 * Copyright (c) 2026 mol* contributors, licensed under MIT, See LICENSE file for more info.
 *
 * @author Alexander Rose <alexander.rose@weirdbyte.de>
 */
import { clearStructureWiggle, setStructureWiggleFromUncertainty } from '../../mol-plugin-state/helpers/structure-wiggle.js';
import { CollapsableControls, PurePluginUIComponent } from '../base.js';
import { Button } from '../controls/common.js';
import { AnimationSvg } from '../controls/icons.js';
export class StructureProceduralAnimationControls extends CollapsableControls {
    defaultState() {
        return {
            isCollapsed: false,
            header: 'Procedural Animation',
            brand: { accent: 'gray', svg: AnimationSvg }
        };
    }
    renderControls() {
        return _jsx(StructureProceduralAnimation, {});
    }
}
class StructureProceduralAnimation extends PurePluginUIComponent {
    constructor() {
        super(...arguments);
        this.state = { busy: false };
    }
    get components() {
        return this.plugin.managers.structure.hierarchy.selection.structures.flatMap(s => s.components);
    }
    async applyUncertaintyWiggle() {
        this.setState({ busy: true });
        try {
            const options = this.plugin.managers.structure.component.state.options;
            await this.plugin.managers.structure.component.setOptions({
                ...options,
                animation: {
                    ...options.animation,
                    wiggleAmplitude: 0,
                    tumbleAmplitude: 0
                }
            });
            await setStructureWiggleFromUncertainty(this.plugin, this.components);
        }
        finally {
            this.setState({ busy: false });
        }
    }
    async applyDynamics() {
        this.setState({ busy: true });
        try {
            const options = this.plugin.managers.structure.component.state.options;
            await this.plugin.managers.structure.component.setOptions({
                ...options,
                animation: {
                    ...options.animation,
                    wiggleSpeed: 7,
                    wiggleAmplitude: 1,
                    wiggleFrequency: 0.2,
                }
            });
            await clearStructureWiggle(this.plugin, this.components);
        }
        finally {
            this.setState({ busy: false });
        }
    }
    async clearWiggle() {
        this.setState({ busy: true });
        try {
            const options = this.plugin.managers.structure.component.state.options;
            await this.plugin.managers.structure.component.setOptions({
                ...options,
                animation: {
                    ...options.animation,
                    wiggleAmplitude: 0,
                    tumbleAmplitude: 0
                }
            });
            await clearStructureWiggle(this.plugin, this.components);
        }
        finally {
            this.setState({ busy: false });
        }
    }
    render() {
        return _jsx(_Fragment, { children: _jsxs("div", { className: 'msp-control-group-wrapper', children: [_jsx("div", { className: 'msp-control-group-header', children: _jsx("div", { children: _jsx("b", { children: "Apply Wiggle" }) }) }), _jsxs("div", { className: 'msp-flex-row', children: [_jsx(Button, { title: 'Set wiggle speed to 8 and amplitude to 1', onClick: () => this.applyDynamics(), disabled: this.state.busy, children: "Dynamics" }), _jsx(Button, { title: 'Set per-group wiggle amplitude based on B-factor / RMSF uncertainty', onClick: () => this.applyUncertaintyWiggle(), disabled: this.state.busy, children: "Uncertainty" }), _jsx(Button, { title: 'Set wiggle/tumble amplitude to zero and remove per-group wiggle layers', onClick: () => this.clearWiggle(), disabled: this.state.busy, children: "Clear" })] })] }) });
    }
}
